src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -437,51 +464,143 @@ public void delete(String modelId, ActionListener<DeleteModelResponse> listener)
+                    String errorMessage = String.format("Cannot delete model \"%s\". Model is still in training", modelId);
+                    logger.error(errorMessage);
+                    listener.onResponse(new DeleteModelResponse(modelId, "failed", errorMessage));
+                    return;
+                }
+
+                // Add modelId to model graveyard until delete model request is processed
+                updateModelGraveyardToDelete(modelId, false, blockModelIdStep);
+            }, listener::onFailure);
+
+            // Remove the metadata asynchronously
+            blockModelIdStep.whenComplete(
+                acknowledgedResponse -> { clearModelMetadata(modelId, clearModelMetadataStep); },
+                listener::onFailure
+            );
+
             // Setup delete model request
             DeleteRequestBuilder deleteRequestBuilder = new DeleteRequestBuilder(client, DeleteAction.INSTANCE, MODEL_INDEX_NAME);
             deleteRequestBuilder.setId(modelId);
             deleteRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
 
-            // On model deletion from the index, remove the model from all nodes' model cache
-            ActionListener<DeleteResponse> onModelDeleteListener = ActionListener.wrap(deleteResponse -> {
-                // If model is not deleted, return with error message
+            // On model metadata removal, delete the model from the index
+            clearModelMetadataStep.whenComplete(
+                acknowledgedResponse -> deleteModelFromIndex(modelId, deleteModelFromIndexStep, deleteRequestBuilder),
+                listener::onFailure
+            );
+
+            deleteModelFromIndexStep.whenComplete(deleteResponse -> {
+                // If model is not deleted, remove modelId from model graveyard and return with error message
                 if (deleteResponse.getResult() != DocWriteResponse.Result.DELETED) {
+                    updateModelGraveyardToDelete(modelId, true, unblockModelIdStep);
                     String errorMessage = String.format("Model \" %s \" does not exist", modelId);
                     listener.onResponse(new DeleteModelResponse(modelId, deleteResponse.getResult().getLowercase(), errorMessage));
                     return;
                 }
 
-                // After model is deleted from the index, make sure the model is evicted from every cache in the
-                // cluster
-                client.execute(
-                    RemoveModelFromCacheAction.INSTANCE,
-                    new RemoveModelFromCacheRequest(modelId),
-                    ActionListener.wrap(removeModelFromCacheResponse -> {
+                // After model is deleted from the index, make sure the model is evicted from every cache in the cluster
+                removeModelFromCache(modelId, clearModelFromCacheStep);
+            }, e -> listener.onResponse(new DeleteModelResponse(modelId, "failed", e.getMessage())));
```

Comment:
If "failed" is something standard/reusable we need to pul it to the class level constant

Reply:
ack